### ISSUE TITLE
CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version (Acme::Client::Error::Malformed)

### ISSUE THREAD

**Issue by jbweston** [will-in-wi/letsencrypt-webfaction#38]
I get the above error when running `letsencrypt-webfaction` version `1.1.4`.
The full traceback follows:
```
/home/my_user/.letsencrypt_webfaction/gems/gems/acme-client-0.3.6/lib/acme/client/faraday_middleware.rb:37:in `raise_on_error!': CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version (Acme::Client::Error::Malformed)
	from /home/my_user/.letsencrypt_webfaction/gems/gems/acme-client-0.3.6/lib/acme/client/faraday_middleware.rb:27:in `on_complete'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/acme-client-0.3.6/lib/acme/client/faraday_middleware.rb:12:in `block in call'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/faraday-0.9.2/lib/faraday/response.rb:57:in `on_complete'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/acme-client-0.3.6/lib/acme/client/faraday_middleware.rb:12:in `call'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:139:in `build_response'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/faraday-0.9.2/lib/faraday/connection.rb:377:in `run_request'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/faraday-0.9.2/lib/faraday/connection.rb:177:in `post'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/acme-client-0.3.6/lib/acme/client.rb:46:in `new_certificate'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/letsencrypt_webfaction-1.1.4/lib/letsencrypt_webfaction/application.rb:51:in `certificate'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/letsencrypt_webfaction-1.1.4/lib/letsencrypt_webfaction/application.rb:44:in `certificate_writer'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/letsencrypt_webfaction-1.1.4/lib/letsencrypt_webfaction/application.rb:27:in `run!'
	from /home/my_user/.letsencrypt_webfaction/gems/gems/letsencrypt_webfaction-1.1.4/exe/letsencrypt_webfaction:5:in `<top (required)>'
	from /home/my_user/.letsencrypt_webfaction/gems/bin/letsencrypt_webfaction:23:in `load'
	from /home/my_user/.letsencrypt_webfaction/gems/bin/letsencrypt_webfaction:23:in `<main>'
```

The error message seems pretty self explanatory, although without digging into the library it's not clear to me whether it's the call into `acme-client` that needs to be modified, or whether this is an upstream bug in `acme-client`.

**Comment by will-in-wi** [comment]
Which CentOS are you using? https://docs.webfaction.com/user-guide/server.html#finding-your-server-s-operating-system

My first guess is that WebFaction is including an older version of OpenSSL.

**Comment by jbweston** [comment]
Ostensibly CentOS 6 (`/etc/redhat-release` reports `CloudLinux Server release 6.8 (Oleg Makarov)`), and `openssl version` reports `OpenSSL 1.0.1e-fips 11 Feb 2013`, (although the build date is from this year, so I guess it's patched for heartbleed).

Are certs generatedby older versions of openssl not supported at all by the letsencrypt API? (a cursory google doesn't reveal anything, and [this post](https://community.letsencrypt.org/t/howto-a-with-all-100-s-on-ssl-labs-test-using-apache2-4-read-warnings/2436) seems to suggest that there is no such requirement).

**Comment by panzerdp** [comment]
Faced now the same problem.
A few days I generated a certificate for other domain and it was working fine, for the same server.

**Comment by will-in-wi** [comment]
This looks like it was introduced upstream in Let's Encrypt itself: https://github.com/letsencrypt/boulder/blob/7e946eaacc30e5625d49a8a5bba597854a52f7bd/wfe/wfe.go#L809

I'm investigating solutions.

**Comment by will-in-wi** [comment]
As a short term "solution", albeit one that is rather complicated, you can use the CentOS 5 instructions where you compile your own OpenSSL: https://github.com/will-in-wi/letsencrypt-webfaction/wiki/Install-custom-OpenSSL-and-Ruby-on-CentOS-5-host

The only alternative I can see is to get Let's Encrypt to remove the restriction or for WebFaction to upgrade OpenSSL. I'm not sure either are likely to happen. If both of you create WebFaction support tickets around this, it might help.

**Comment by will-in-wi** [comment]
I may have a better solution, involving patching and releasing ACME::Client. If you can hold out for a few days while I get this working, feel free. Otherwise, the link above should provide a short term fix.

**Comment by panzerdp** [comment]
I downloaded and compiled a new version of `openssl` 1.0.2 using [these](https://community.webfaction.com/questions/7959/how-did-i-install-my-own-phpcurlopenssl-for-use-with-the-shared-web-server) instructions.
However I do receive the same [error message](http://dl.dropbox.com/u/93627719/315077.png).
Weird, just a few days ago it was working fine without any custom configurations.

**Comment by will-in-wi** [comment]
You would have to recompile Ruby to use the new OpenSSL.

**Comment by will-in-wi** [comment]
Let's Encrypt added a new restriction, which is why it suddenly stopped working. I think I have a fix, I'm working on testing it.

**Comment by will-in-wi** [comment]
You can see some of the technical details here: https://github.com/unixcharles/acme-client/issues/84

**Comment by jbweston** [comment]
appreciate the rapid response! :+1:

**Comment by panzerdp** [comment]
Thanks @will-in-wi for your support!
When you have a version, let me know I will test.

**Comment by will-in-wi** [comment]
No problem. This'll break my utility for possibly the majority of users, so I want to get it figured out ASAP.

**Comment by will-in-wi** [comment]
Okay, I confirmed that https://github.com/unixcharles/acme-client/pull/85 fixes this. I've submitted the upstream patch. As soon as @unixcharles merges and releases a new version, I'll cut a new version which requires the new version of acme-client.

Sorry for the issue. Turns out that my personal WF install has the same issue, so thanks for reporting before all of my certs simultaneously failed to renew.

**Comment by panzerdp** [comment]
@will-in-wi Not sure if viable, but for these situations it's nice to have an email notification that a certificate failed to generate. 
A new command argument can be created `--notify-on-fail`. It can be used for the cron command: `letsencrypt_webfaction --notify-on-fail [....]`. When cron runs the command and it fails, user is notified about the situation with an email.
At the moment I think it fails silently.

**Comment by will-in-wi** [comment]
I've handled this in the past by adding `MAILTO=myemail@mydomain.com` to the top of the crontab. This emails me whenever any cron task has standard out or standard error. It's generally good practice so that you know when cron jobs fail.

You can test this by adding the aformentioned line and then adding `* * * * *   echo "Here's an email!"` to your crontab. You should start receiving an email every minute.

**Comment by will-in-wi** [comment]
Does this solution work for you? I'm open to adding the suggested flag, I just don't want to do so if cron includes this feature for free.

**Comment by panzerdp** [comment]
@will-in-wi Sounds like a good solution, thanks for the tip. (Maybe worth mentioning in the docs)

**Comment by will-in-wi** [comment]
Good point. Added: https://github.com/will-in-wi/letsencrypt-webfaction/commit/6fce9526fb204440892942ac38ac7200750a8bf1

**Comment by will-in-wi** [comment]
@panzerdp @jbweston: Let's Encrypt wants to know if you are seeing this on the production instance or just the staging instance. E.g. when you saw this issue, were you using the `--endpoint https://acme-staging.api.letsencrypt.org/` parameter?

https://github.com/letsencrypt/boulder/issues/1902#issuecomment-233032593

I'll relay your responses, or you can comment directly on the ticket.

**Comment by will-in-wi** [comment]
It sounds like this upstream change only affects staging. Acme::Client accepted my PR to fix this when the LE change does actually hit production.

So, if you want to just go to prod, the default endpoint should work fine. I'll continue to get the fix out the door so that we aren't hit with a breakage when they move the change to production.

**Comment by panzerdp** [comment]
@will-in-wi 
I just used the command for production and it worked fine. It seems like it fails for staging only. 
Thank you.

**Comment by nmcbride** [comment]
I'm going to subscribe to this thread. When you push the fix can you comment here? I'm really bad at tracking things like this. XD

**Comment by will-in-wi** [comment]
I just released version 0.1.5, which requires acme-client `>= 0.3.7`. This fixes the openssl issue.
